### ISSUE TITLE
Bug in the PersistenceUnitProcessor.findPersistenceArchives(ClassLoader loader, String descriptorPath)

### ISSUE THREAD

**Issue by ah-ping-luk** [eclipse-ee4j/eclipselink#1348]
I hit an issue when i pack an app in WAR with structure:
- /WEB-INF/classes/META-INF/persistence.xml

inside the PersistenceUnitProcessor.findPersistenceArchives(ClassLoader loader, String descriptorPath):

//where the puRootUrl = jar:file:/..../WEB-INF/classes/' and the descriptorPath ( = META-INF/persistence.xml [the default location])

archive = PersistenceUnitProcessor.getArchiveFactory(loader).createArchive(puRootUrl, descriptorPath, null);

where the archive object will end-up be an 'DirectoryInsideJarURLArchive' instance, with method:

 //entryPath = 'META-INF/persistence.xml'
    public InputStream getEntry(String entryPath) throws IOException {
        InputStream is = entries.contains(entryPath) ?
            jarFile.getInputStream(jarFile.getEntry(relativeRootPath + entryPath)) : null;
        return is;
    }

//the entries.contains(entryPath) will always return false (as the entries contains strings with leading '/' but the entryPath isn't)

also the (relativeRootPath + entryPath) should be => (relativeRootPath + "/" + entryPath) 
//as the relativeRootPath = 'WEB-INF/classes' which form a path string '**WEB-INF/classesMETA-INF/persistence.xml**'

---------------------------

This bug which finally lead to NPE and not able to load the persistence.xml file correctly.

is it a bug in the PersistenceUnitProcessor or the DirectoryInsideJarURLArchive class?

---------------------------

source references:
- https://github.com/eclipse-ee4j/eclipselink/blob/master/jpa/org.eclipse.persistence.jpa/src/main/java/org/eclipse/persistence/internal/jpa/deployment/DirectoryInsideJarURLArchive.java
- https://github.com/eclipse-ee4j/eclipselink/blob/master/jpa/org.eclipse.persistence.jpa/src/main/java/org/eclipse/persistence/internal/jpa/deployment/PersistenceUnitProcessor.java
-

**Comment by rfelcman** [comment]
Could You please describe context where this bug happens? I didn't saw issue bug like this during deployment or web application call (packaged in WAR).

**Comment by ah-ping-luk** [comment]
I hit this issue when i try to migrate a legacy application (EJB 3.x + JSP/Servlet) to springboot application.

To support JSP in springboot, i need to pack the application in .war format but run it as a stanalone app.

The legacy application do make use of programmatic persistence for connecting to different DBs (based on different request parameters) (i.e. we do create the EntityManager/EntityManagerFactory manually).

Sample workaround i used below when creating the EntityManagerFactory:
```
HashMap<String, String> h = new HashMap<String, String>();
h.put(PersistenceUnitProperties.JTA_DATASOURCE, ds);

//the default location of persistence.xml
String puFileLocation = "META-INF/my-persistence.xml";

String file = Thread.currentThread().getContextClassLoader().getResource(puFileLocation).toString();

if(file.startsWith("jar:")) { //e.g. check if file = jar:file:/abc.war/WEB-INF/classes/META-INF/persistence.xml 
	//there is a bug in eclipselink for loading the correct path of PU in WAR file in springboot
	
	puFileLocation = "/" + puFileLocation;	
}
			
h.put(PersistenceUnitProperties.ECLIPSELINK_PERSISTENCE_XML, puFileLocation);

EntityManagerFactory emf = Persistence.createEntityManagerFactory(persistenceUnit, h);

```

**Comment by rfelcman** [comment]
I fixed this issue in master (4.0) and 3.0 branches (booth are jakarta.persistence package implementations ) but I don't see problem with 2.7 branch following patterns work there:
With system property
``
System.setProperty(PersistenceUnitProperties.ECLIPSELINK_PERSISTENCE_XML, "WEB-INF/classes/META-INF/persistence.xml");
EntityManagerFactory entityManagerFactory = Persistence.createEntityManagerFactory("test-jpa-pu2");
EntityManager em = entityManagerFactory.createEntityManager();
``

With PersistenceUnit property
``
Properties persistenceUnitProperties = new Properties();
persistenceUnitProperties.setProperty(PersistenceUnitProperties.ECLIPSELINK_PERSISTENCE_XML, "WEB-INF/classes/META-INF/persistence.xml");
EntityManagerFactory entityManagerFactory = Persistence.createEntityManagerFactory("test-jpa-pu2", persistenceUnitProperties);
``

Which version of the EclipseLink did You mention in the bug?
